### PR TITLE
Add quickstart tutorial.

### PR DESCRIPTION
**Pull request by jzohrab**
Compilation of information from several sources into a single canonical document for new users.

I added a separate Vagrantfile, rather than reusing the one at test/unit/Vagrant/Vagrantfile, as I wanted to keep the vagrantfile short focussed, and to reduce the version numbering issues (the new Vagrantfile hardcodes the box name as "vEOS-lab-quickstart" per the tutorial instructions).

Addresses https://github.com/napalm-automation/napalm/issues/230

### CODE REVIEW

File: docs/tutorials/quickstart.rst [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+We'll use Arista, which can be downloaded for free from the Arista site.
+
+Create an account at https://www.arista.com/en/user-registration, and go to https://www.arista.com/en/support/software-download.
+
+Download the latest "vEOS-lab-<version>-virtualbox.box" listed in the vEOS folder at the bottom of the page.
+
+Add it to your vagrant box list as "vEOS-lab-quickstart"::
+
+    $ vagrant box add --name vEOS-lab-quickstart ~/Downloads/vEOS-lab-<version>-virtualbox.box
```
Comment (dbarrosop) [inline docs/tutorials/quickstart.rst]: I would use their generic name `vEOS-lab-<version>-virtualbox` instead of a specific one. It's supposed to be portable... : )

**Comment by dbarrosop** [comment]
I really like it, I think a few things might be interesting as well:

1. Change the current tutorials to use the same vagrantfile you have. With that I mean: http://napalm.readthedocs.org/en/latest/tutorials/index.html and http://napalm.readthedocs.org/en/latest/cli.html
1. Reference all the tutorials in your document. I think it's interesting that after seeing how simple the cli is they can jump into some python code and see that's it's almost as simple and not that different from using just the CLIs they are used to... Just a quick introduction to python code and then a link to the longer tutorial in the same way you did the `cli` introduction.

Does it make sense?

File: docs/tutorials/quickstart.rst [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+Installation
+------------
+
+If you haven't already installed NAPALM, you can run the code from a virtual environment in the root directory::
+
+    $ virtualenv venv                        # create environment
+    $ source venv/bin/activate               # start environment
+    $ sudo pip install -r requirements.txt   # install dependencies
+    $ python setup.py install                # install NAPALM
```
Comment (dbarrosop) [inline docs/tutorials/quickstart.rst]: Don't use this method. Use pip instead, please. There are some caveats when installing the new modular code with this mechanism.

File: docs/tutorials/quickstart.rst [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+Notes:
+
+* we're setting the ``--name`` parameter here so that the tutorial's Vagrantfile will work out-of-the-box.  You may wish to also ``vagrant box add`` this file with the correct version
+* ``vagrant box add`` copies downloaded file to a designated directory (e.g., for Mac OS X and Linux: ``~/.vagrant.d/boxes``, Windows: ``C:/Users/USERNAME/.vagrant.d/boxes``).
+
+Starting Vagrant
+----------------
+
+The Vagrantfile (in this directory) creates a base box and a vEOS box when you call "vagrant up"::
```
Comment (dbarrosop) [inline docs/tutorials/quickstart.rst]: As pip is the preferred method to install napalm I guess that you should mention first that either they download the Vagrantfile or they just create the file and copy & paste the content. I guess downloading is better to avoid issues.

File: docs/tutorials/quickstart.rst (hunk 1 above)
Comment (jzohrab) [inline docs/tutorials/quickstart.rst]: Yes, the reason for setting the `--name` like this is so that the vagrantfile can find the box, without them changing the vagrantfile.  The file name is still `...<version>..`.  I specifically didn't want to say "update the Vagrantfile to use the same version you downloaded earlier".

File: docs/tutorials/quickstart.rst (hunk 3 above)
Comment (jzohrab) [inline docs/tutorials/quickstart.rst]: Ah right, I was thinking that they'd clone this repo, which is likely too much for some people.  Got it, thanks.

File: docs/tutorials/quickstart.rst (hunk 2 above)
Comment (jzohrab) [inline docs/tutorials/quickstart.rst]: I see.  I was thinking about this from the viewpoint of a dev who'd checked everything out.  So should this be `sudo pip install napalm`?

**Comment by jzohrab** [comment]
@dbarrosop - re "I think a few things might be interesting as well" - yes, I agree.  How about we work on that as separate issues once we work out the points you're raised, and this gets into master?  This keeps this PR's scope focused.  I agree with what you say, it does provide a nice starting point for those tutorials and would provide something concrete to work with.  Presumably, "set up your lab" would be split out into a separate section entirely, and referred to from those tutorials.

File: docs/tutorials/quickstart.rst (hunk 1 above)
Comment (dbarrosop) [inline docs/tutorials/quickstart.rst]: What I meant is that I think it's better if we stick with the default name on the Vagrantfile. Maybe some people already have it or might start using it for other purposes.

File: docs/tutorials/quickstart.rst (hunk 2 above)
Comment (dbarrosop) [inline docs/tutorials/quickstart.rst]: From a developer's point of view is a bit more complex than that. Wait until the refactoring is merged because when it's done setting up the dev environment will involve cloning a few repos and tweaking the pythonpath. Don't worry, is not as hard as it sounds and it can be easily automated ;)

BTW, you can always do `pip install -e .` if you are inside the folder to install the cloned repository with pip.

File: docs/tutorials/quickstart.rst (hunk 1 above)
Comment (jzohrab) [inline docs/tutorials/quickstart.rst]: Right, I see.  Since we're talking about people installing with pip and copying/downloading the vagrantfile, rather than doing a `git checkout` as I'd originally thought, then using the name as-is makes sense.  If it was strict version control, I'd say use the name as suggested - but that implies that devs would be using this, and they should know enough anyway.  Sounds good, will change.

File: docs/tutorials/quickstart.rst (hunk 2 above)
Comment (jzohrab) [inline docs/tutorials/quickstart.rst]: ok, np.  I see the instructions on the README, will change.  Thanks.

**Comment by jzohrab** [comment]
Will make the changes suggested, repush the branch, and then will ping you to review again.  I won't rebase and squash the branch unless you'd prefer that for git commit history.  Thanks for all the feedback, appreciated.

**Comment by jzohrab** [comment]
Making some changes and repushing.

**Comment by jzohrab** [comment]
@dbarrosop - I committed some changes.  the docs look OK to me when generated with sphinx locally.  Once this gets checked and  into master, we can start to iterate on the other suggestions you had.

File: docs/tutorials/quickstart.rst [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    $ cl_napalm_configure \
+      --user vagrant \
+      --vendor eos \
+      --strategy replace \
+      --optional_args 'port=12443' \
+      --dry-run \
+      new_good.conf localhost
+
+
+Try the other examples and commands given at http://napalm.readthedocs.org/en/latest/cli.html.
```
Comment (dbarrosop) [inline docs/tutorials/quickstart.rst]: Maybe we should add instructions on the cli page to download the rest of the files as you did here. Otherwise, going from here to the link and keep following it's going to result in failures and frustrations : )

File: docs/tutorials/quickstart.rst [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    $ cd PROJECT_ROOT/docs/tutorials
+    $ vagrant destroy -f
+    $ deactivate           # exit the virtualenv environment
+
+Next Steps
+----------
+
+There are many possible steps you could take next:
+
+* create Vagrant boxes for other devices
```
Comment (dbarrosop) [inline docs/tutorials/quickstart.rst]: Maybe you can link to instructions/examples on what you mean with that? For example, I wrote this post some time ago https://www.dravetech.com/blog/2016/01/14/vagrant_box_ios_xr.html. However, you can link something else if you know of some other resource.

File: docs/tutorials/quickstart.rst [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+* explore using configuration management tools (Ansible, Chef, Salt, etc.)
```
Comment (dbarrosop) [inline docs/tutorials/quickstart.rst]: Same here, some links might be interesting.

File: docs/tutorials/quickstart.rst [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+      --dry-run \
+      new_good.conf localhost
+
+
+Try the other examples and commands given at http://napalm.readthedocs.org/en/latest/cli.html.
+
+Shutting down
+-------------
+
+    $ cd PROJECT_ROOT/docs/tutorials
```
Comment (dbarrosop) [inline docs/tutorials/quickstart.rst]: This is the first time this path appears, I guess it's some leftover. By the way, this gave me an idea. Do you want me to create a `napalm-tutorials` repo where you can put the files needed to run this and other tutorials? No documentation there, just files so you can start with:

```
git clone https://github.com/napalm-automation/napalm-tutorials
cd napalm-tutorials/quickstart```

and take it from there.

File: docs/tutorials/quickstart.rst (hunk 4 above)
Comment (jzohrab) [inline docs/tutorials/quickstart.rst]: Reflecting further on test files: I think it would be best to actually copy the test files from the eos repo into this repo, and to use them here.  The files are small, and can be made even smaller.  Then all tutorials could reference the same files, and they could all be updated.  Let's do that in a separate PR.

File: docs/tutorials/quickstart.rst (hunk 7 above)
Comment (jzohrab) [inline docs/tutorials/quickstart.rst]: Right, I needed a way for people to get back to the folder with the Vagrantfile, typing "vagrant destroy" in any other directory doesn't work.  This was the cleanest way I could think of to write it.

I would recommend keeping the tutorial in alongside the rest of the code, rather than as a separate repo.  Ideally, the tutorial instructions should be tested alongside the code, to ensure that the instructions stay current.  By "testing" here, I mean actual automated testing, and not relying on people to remember to test things when they change code.  This isn't actually done here yet, we could backlog that idea.

@dbarrosop - can we merge this into master?  I have two reasons for the request:

1) It looks like we're talking about improvements that we can make iteratively.  "Release early, release often" is a good thing to follow, sometimes.  Also, once it's in, everyone can make contributions to it and it could go faster.
2) we're talking about changing some of the other tutorials, and I don't want the scope to creep.  Small PRs are good PRs!
3) A colleague was asking for this quickstart material, if this gets into master and onto "readthedocs" it will help our team. :-)

If you would rather keep this separate, I can continue to iterate on this, but I'd prefer to merge it and add issues that I'll deal with quickly in separate smaller PRs.

Comment (dbarrosop) [inline docs/tutorials/quickstart.rst]: WALL OF TEXT ALERT!!! : )

Honestly, I would rather not merge yet. I see the value of what we are trying to do but I don't think we are there yet. I also think there is some confusion here. But that might be due to the latest changes in NAPALM or due to the lack of "common goals" for this PR.

First, some notes and clarifications regarding napalm.

> I would recommend keeping the tutorial in alongside the rest of the code, rather than as a separate repo

The repo `napalm` contains no code. It's only a placeholder for documentation and a wrapper to install all the drivers. The common code of napalm lives now in `napalm-base` and each driver has a dedicated repo, ie, `napalm-eos`, `napalm-ios`, etc... There was some discussion in the mailing list about this a month or two ago and the latest release notes has also some explanations around this and why we did it. In any case, what I meant by having a dedicated repo where the support files, not the content. However, I don't mind having the support files in the main `napalm` repo either, at the end of the day, this is one of the main reasons of this repo.

> the tutorial instructions should be tested alongside the code, to ensure that the instructions stay current. By "testing" here, I mean actual automated testing, and not relying on people to remember to test things when they change code. This isn't actually done here yet, we could backlog that idea.

That is not necessary. Two reasons:

1. The API is stable, that means that code written following napalm API should work, always, no matter what.
2. Each driver has it's own set of integration tests that are run automatically every time someone pushes new code. So we should find bugs and issues there without the need of testing the quickstart. I am not claiming the tests are perfect so feel free to comment if you think something could be improved. Bottom line here is that we already have automated testing covering this.

> "Release early, release often" is a good thing to follow, sometimes.

Agree, that was mostly the reason behind the latest refactoring. That and smaller PRs. However, features should be complete, releasing something that doesn't work might just frustrate people and make them lose confidence on the project so I am all for releasing quick and fast, as long as is complete. Specially if it's documentation. A common concern on OSS is that documentation is confusing or that you can't just follow it because you always find surprises.

> A colleague was asking for this quickstart material, if this gets into master and onto "readthedocs" it will help our team. :-)

But this is not about one particular's use case. It's about the community. Being a contributor has the advantage that you can focus on your particular problem but we should still aim to build something as generic as possible so everybody could get the benefit of it. I know that's probably not what you meant but I felt obliged to clarify anyway ; )

Back to the PR itself. Right now, I think that most of the content of this PR mostly duplicates work already done in the tutorials. So let me try to explain what I envision for this PR:

1. The target audience should be developers/network_engineers/users_in_general that heard about napalm and what to see it in action and maybe try it by themselves. The key word here is **users**. Anyone that might be considering napalm as their main interface to the network. They could be people looking into ansible, writing their own scripts or something else. But definitively not people trying to contribute to napalm. For that people we will have some other document.
2. The goal of the document should be to be able to copy & paste a few lines and see different things you can do with napalm:
   1. Change configuration
   1. Get diffs.
   1. Get LLDP neighbors/BGP/interfaces/whatever
   1. Test the CLI tool. 
3. Provide a list of resources to explore a bit further.

Right now, most of the stuff is covered on the tutorials, what is missing is:

1. How to quickly build a lab so you don't have to test on your own network.
2. Fixing the tutorials so you can actually copy & paste them.

In summary, my problem with the PR right now is that overlaps a lot with the tutorials. I would rather have this PR just telling you how to start a lab to move forward onto the rest of the tutorials and adapt the tutorials so they can be reproduced with the lab.

I hope it makes sense, otherwise, feel free to send me back a wall of text :P

**Comment by jzohrab** [comment]
Closing in favor of PR #238 , we'll iterate on that branch.  Thanks @dbarrosop for the feedback.
